### ISSUE TITLE
enums and redefined-variable-type

### ISSUE THREAD

**Issue by ETayrienHBO** [PyCQA/pylint#777]
I'm using python 3.4.2 and pylint 1.5.2. Pylint is printing a r:redefined-variable-type warning when I use a variable of type enum. The following script demonstrates the problem:

```
from enum import Enum

class MyEnum(Enum):
    Value1 = 1
    Value2 = 2

myEnum = MyEnum.Value1
print("my enum is {}".format(myEnum))

# Huh? this line generates
# warning R0204: Redefinition of myEnum type from EnumRepro.MyEnum.Value1 to 
# EnumRepro.MyEnum.Value2 [R:redefined-variable-type]
myEnum = MyEnum.Value2 # this generates 
print("my enum is {}".format(myEnum))
```

Is this expected? At the moment I’m just disabling R0204 at the code locations that produce the problem, which isn't great.

mentioning @lmedioni per suggestion from mailing list. This issue is potentially related to 
https://github.com/PyCQA/pylint/issues/710

**Comment by sigmavirus24** [comment]
I'm having the same problem. For a slightly different explanation, I have some code that looks roughly like

```py
import enum

class Decision(enum.Enum):
     Allowed = 'allowed'
     Rejected = 'rejected'
     Pending = 'pending'


def explicitly_allowed(*inputs):
      # returns either True or False


def explicitly_rejected(*inputs):
      # returns either True or False

class MyClass(object):
    def __init__(self):
        self._decision_cache = {}

    def decide(*inputs):
        decision = self._decision_cache.get(inputs)
        if decision is not None:
              return decision
        if explicitly_allowed(*inputs) and not explicitly_rejected(*inputs):
              decision = Decision.Allowed
        elif explicitly_rejected(*inputs) and not explicitly_allowed(*inputs):
              decision = Decision.Rejected
        else:
              decision = Decision.Pending
        self._decision_cache[inputs] = decision
        return decision
```

Since we're dealing with an Enum, I would expect that each assignment to `decision` would be of the same type (`Decision`). Let me know if there's a way I can help out with this. I haven't contributed to pylint before but I'd be happy to give it a shot with this.

**Comment by PCManticore** [comment]
Hey Ian,

Sorry for not replying earlier, I was sick and couldn't be in front of a computer.

I'd definitely need some help with this pattern, unfortunately it is very buggy with regard to false positives and the fact that I didn't receive almost any help from the original authors made me to move to the extension status in 2.0, until it can be reliable again. This means that in order to get the error again in the master branch, you'll have to run ``pylint a.py --load-plugins=pylint.extensions.check_reassign`` in order to enable it again (the master of pylint works only with the master of astroid for now)

The main problem of this pattern is that it relies on the string representation of the types, which are usually not what we want and it's a poor excuse for not using structural subtyping. This means that a change to this pattern can imply verifying the structure of the types involved, if they seem to have the same structure, then they can be used in the same contexts through duck typing. For instance, you might see this example in https://github.com/PyCQA/pylint/issues/772. You can try to see if those types are similar by looking at their methods and if they seem to have the same method, then it's good to go. This is not ideal, since some of these methods can do different things for each type, but we'll have to leave with it until we'll have PEP 484 support.

Let me know if you need any help, I can jump in on IRC.

**Comment by PCManticore** [comment]
Closing this. redefined-variable-type is now an extension in pylint 1.7 and we will probably not spend more effort improving this check.
